tests/test_connection.py

```diff
@@ -341,3 +349,274 @@ def cb1(n):
+
+    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+    assert cur.execute("select * from temp_table").fetchone() == ("foo",)
+
+
+def test_transaction_rollback_on_exception_exit(temp_table, conn):
+    """Changes are rolled back if an exception escapes the `with` block."""
+    cur = conn.cursor()
+    with pytest.raises(ExpectedException):
+        with conn.transaction():
+            cur.execute("insert into temp_table values ('foo')")
+            raise ExpectedException("This discards the insert")
+
+    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+    assert cur.execute("select * from temp_table").fetchone() is None
+
+
+def test_transaction_prohibits_use_of_commit_rollback_autocommit(conn):
+    """
+    Within a Transaction block, it is forbidden to touch commit, rollback,
+    or the autocommit setting on the connection, as this would interfere
+    with the transaction scope being managed by the Transaction block.
+    """
+    conn.autocommit = False
+    conn.commit()
+    conn.rollback()
+
+    with conn.transaction():
+        message = re.escape(
+            "can't change autocommit state when in Transaction context"
+        )
+        with pytest.raises(psycopg3.ProgrammingError, match=f"^{message}$"):
+            conn.autocommit = False
+
+        message = re.escape(
+            "Explicit commit() forbidden within a Transaction context."
+            " (Transaction will be automatically committed on successful exit"
+            " from context.)"
+        )
+        with pytest.raises(psycopg3.ProgrammingError, match=f"^{message}$"):
+            conn.commit()
+
+        message = re.escape(
+            "Explicit rollback() forbidden within a Transaction context. "
+            "(Either raise Transaction.Rollback() or allow an exception to "
+            "propagate out of the context.)"
+        )
+        with pytest.raises(psycopg3.ProgrammingError, match=f"^{message}$"):
+            conn.rollback()
+
+    conn.autocommit = False
+    conn.commit()
+    conn.rollback()
+
+
+@contextmanager
+def _transaction(conn, exception):
+    """
+    Context manager does things in a conn.transaction() and then optionally
+    exits the context manager with an exception.
+    """
+    with pytest.raises(ExpectedException) if exception else nullcontext():
+        with conn.transaction():
+            yield
+            if exception:
+                raise ExpectedException()
+
+
+@pytest.mark.parametrize("exception", [False, True])
+@pytest.mark.parametrize("autocommit", [False, True])
+def test_transaction_preserves_autocommit(conn, autocommit, exception):
+    """
+    Connection.autocommit value is False in Transaction block, but the original
+    value is always restored after the block exits, both in successful exit and
+    exception scenarios.
+    """
+    conn.autocommit = autocommit
+    with _transaction(conn, exception):
+        assert conn.autocommit is False
```

Comment:
I really don't like this behaviour. This is an user choice, it shouldn't be messed with

Reply:
Interesting thought. Maybe this is an unnecessary hangover from my previous implementation, where autocommit was the only way to control transactions so you _had_ to mess with it...

So do you think it's acceptable for autocommit to remain `False` inside a transaction block? To me this basically means "any change you make is getting committed immediately and unconditionally" which is not true inside a transaction block, but I'm also happy to say "if you're using transaction blocks, don't look at autocommit!"

Reply:
If you are in autocommit and people do `cur.execute("begin")` you are in a transaction too. `autocommit` only says that the connection won't start a transaction implicitly. If you start it explicitly (with a statement, or with `conn.transaction()`) it's not less-auto, it's pretty manual.